README.md
# EBANX Payment Gateway for WooCommerce

[![Build Status](https://travis-ci.com/ebanx/woocommerce-gateway-ebanx.svg?branch=master)](https://travis-ci.com/ebanx/woocommerce-gateway-ebanx)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/09ef5eb63a394dc3b76cb4319129fbf3)](https://www.codacy.com/app/EBANX/woocommerce-gateway-ebanx?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ebanx/woocommerce-gateway-ebanx&amp;utm_campaign=Badge_Grade)

Integrate your WooCommerce store with the EBANX payment gateway and improve your sales in fastest growing markets in Latin America.

Please, visit the [official plugin page on WordPress store](https://wordpress.org/plugins/ebanx-payment-gateway-for-woocommerce/).

## Requirements
* PHP 5.6+
* Wordpress 4.0+
* WooCommerce 3.0+

## Installation

You can use our docker.

First install dependencies:

```ssh
composer install
```

Then add to `/etc/hosts` file:

```ssh
127.0.0.1 wclocal
```

And run docker compose:

```ssh
docker-compose up
```

Access in your browser: http://localhost

### Do you want to contribute?

We've put together all the information you'll need on [CONTRIBUTING.md](https://github.com/ebanx/woocommerce-gateway-ebanx/blob/master/CONTRIBUTING.md). Also, we always try to keep this file updated and with as much information as possible. If you want to help us with this file or with any coding issue, we would appreciate it. :heart:

### Are you a Merchant?

We have three ways to install our plugin. Choose one and install it. ;)

#### Automatic

Automatic installation is the easiest, and it can be done without leaving your web browser. To do an automatic install of the EBANX plugin, login to the WordPress Dashboard, go to the Plugins menu, and select “Add New.” Then, search for the “EBANX” to find EBANX Payment Gateway for WooCommerce and click on “Install Now.” After the installation, you just have to activate and get ready to sell even more.

#### Manual

To install the plugin manually, download our plugin and upload it to your web server via an FTP application. Visit the [WordPress codex](https://codex.wordpress.org/Managing_Plugins#Manual_Plugin_Installation "WordPress codex") for further instructions. Don't waste your time and improve your results!

#### Via GitHub

The EBANX Gateway Plugin can also be installed using GitHub. You can find our repository [here](https://github.com/ebanx/checkout-woocommerce/ "EBANX GitHub repository"). To download the plugin from our repository, please view [our latest release](https://github.com/ebanx/woocommerce-gateway-ebanx/releases/latest "Latest Release from GitHub repository") and download the `woocommerce-gateway-ebanx.zip` package.

### Don't have an EBANX Account yet?

No problem! Access our [sign-up](https://www.ebanx.com/business/en/dashboard "EBANX Dashboard") or [request a call](https://business.ebanx.com/en/contact-sales) with a Business Development Executive.

---

Still need your API Keys? You can pick those up [here](https://www.ebanx.com/business/en/dashboard "EBANX API keys").

## License

Copyright 2017 EBANX Payments

Licensed under the Apache License, Version 2.0 (the "License");
you may not use these files except in compliance with the License.
You may obtain a copy of the License at

   [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



CHANGELOG.md
# CHANGELOG

## 3.4.6
* 2021-03-24 - Disable installments option in change payment method page.

## 3.4.5
* 2021-02-05 - Search metadata in parent order when not found on subscription.

## 3.4.4
* 2021-01-26 - Limit the search of metadata to the last 12 orders and performance adjustments in the subscription renewal.

## 3.4.3
* 2021-01-25 - Update subscription metadata in the renewal.

## 3.4.2
* 2021-01-22 - Different url for payment detail when in sandbox.

## 3.4.1
* 2021-01-15 - Update Benjamin dependency.

## 3.4.0
* 2021-01-08 - Add support to admin change payment method of subscription.

## 3.3.7
* 2021-01-08 - Fix gateway availability check.

## 3.3.6
* 2021-01-04 - Get metadata from order when not found on subscription.

## 3.3.5
* 2021-01-04 - Save card metadata on subscription instead of saving in the order.

## 3.3.4
* 2020-12-21 - Improve notes on subscription payment process.

## 3.3.3
* 2020-11-03 - Improves and simplifies the filter to handle error messages.

## 3.3.2
* 2020-11-03 - Improve customized error code constant.

## 3.3.1
* 2020-10-30 - Refactor filter to customize API response errors.

## 3.3.0
* 2020-10-30 - Add filter to customize API response errors.

## 3.2.0
* 2020-09-23 - Add support to change payment method by user when order is subscriptions. Associate credit card token on subscription.

## 3.1.1
* 2020-09-09 - Cart with only virtual and downloadable does not require processing.

## 3.1.0
* 2020-09-09 - Add support to create payments with split rules.

## 3.0.4
* 2020-08-31 - Include WooCommerce compatibility version. Fix credit card token resolver.

## 3.0.3
* 2020-06-05 - Fix document fields if is a manual order.

## 3.0.2
* 2020-06-05 - Fix credit card callback when fetching the deviceId and throw the error.

## 3.0.1
* 2020-04-29 - Fix get undefined web service version.

## 3.0.0
* 2020-04-27 - Add support to WooCommerce 4.x. IMPORTANT NOTE: tested on PHP 7.1+, WordPress 5.4+ and WooCommerce 4.0+.

## 2.2.4
* 2020-03-25 - Fix get country error with benjamin update.

## 2.2.3
* 2020-03-25 - Fix checkout with international credit card with foreign document country.

## 2.2.2
* 2020-03-24 - Fix condition to enable/disable instalments on checkout.

## 2.2.1
* 2020-03-18 - Add EBANX javascript lib in global assets.

## 2.2.0
* 2020-03-18 - Enable installment when the cart has subscription.

## 2.1.8
* 2020-03-10 - Add customer document on admin panel. Add instalments field on create new order.

## 2.1.7
* 2020-02-28 - Adds payment related information on admin panel. Adjust translations.

## 2.1.6
* 2020-02-19 - Fix foreign document requirement when country is BR.

## 2.1.5
* 2020-02-19 - Fix foreign document requirement.

## 2.1.4
* 2020-02-19 - Add device finger print on benjamin request. Fix document country for foreign customers.

## 2.1.3
* 2020-02-17 - Fix style to hide document field.

## 2.1.2
* 2020-02-17 - Fix address validation when payment method is international credit card. Fix error when submit form with blank card.

## 2.1.1
* 2020-02-13 - Fix parsing the street number of the address

## 2.1.0
* 2020-02-12 - Add international credit card gateway

## 2.0.14
* 2020-02-08 - Fixes subscriptions cancelling after failed payment

## 2.0.13
* 2020-02-08 - Fixes error when order total is 0 and save user meta fields

## 2.0.12
* 2020-02-07 - Update benjamin library

## 2.0.11
* 2020-01-24 - Fixes error when save user meta fields

## 2.0.10
* 2020-01-24 - Refactor on-click payment process. Add streetComplement on address transformation.

## 2.0.9
* 2020-01-10 - Fixes recurrent payment adding payment_type_code on credit card transaction

## 2.0.8
* 2020-01-10 - Fixes recurrent payment.

## 2.0.7
* 2020-01-10 - Increase plugin version.

## 2.0.6
* 2020-01-10 - Fixes one-click purchase.

## 2.0.5
* 2020-01-10 - Fixes and improve queries. Fixes on get database version.

## 2.0.4
* 2020-01-09 - Fix Text Domain information.

## 2.0.3
* 2020-01-09 - Add billing phone requirement config. Fix optional fields with optional label. Improve translation files.

## 2.0.2
* 2020-01-08 - Improve reademe.txt file.

## 2.0.0
* 2020-01-06 - First Release.


entrypoint.sh
#!/bin/bash

set -euo pipefail

/usr/local/bin/wait-for-it.sh -t 60 mysql:3306 -- echo 'MySQL is up!'

export WP_ROOT=/var/www/html

cd $WP_ROOT

if [ ! -e index.php ] && [ ! -e wp-includes/version.php ]; then
	# if the directory exists and WordPress doesn't appear to be installed AND the permissions of it are root:root, let's chown it (likely a Docker-created directory)
	if [ "$(id -u)" = '0' ] && [ "$(stat -c '%u:%g' .)" = '0:0' ]; then
		chown "www-data:www-data" .
	fi

	echo >&2 "WordPress not found in $PWD - copying now..."
	if [ -n "$(find -mindepth 1 -maxdepth 1 -not -name wp-content)" ]; then
		echo >&2 "WARNING: $PWD is not empty! (copying anyhow)"
	fi
	sourceTarArgs=(
		--create
		--file -
		--directory /usr/src/wordpress
		--owner "www-data" --group "www-data"
	)
	targetTarArgs=(
		--extract
		--file -
	)
	if [ "www-data" != '0' ]; then
		# avoid "tar: .: Cannot utime: Operation not permitted" and "tar: .: Cannot change mode to rwxr-xr-x: Operation not permitted"
		targetTarArgs+=( --no-overwrite-dir )
	fi
	# loop over "pluggable" content in the source, and if it already exists in the destination, skip it
	# https://github.com/docker-library/wordpress/issues/506 ("wp-content" persisted, "akismet" updated, WordPress container restarted/recreated, "akismet" downgraded)
	for contentDir in /usr/src/wordpress/wp-content/*/*/; do
		contentDir="${contentDir%/}"
		[ -d "$contentDir" ] || continue
		contentPath="${contentDir#/usr/src/wordpress/}" # "wp-content/plugins/akismet", etc.
		if [ -d "$PWD/$contentPath" ]; then
			echo >&2 "WARNING: '$PWD/$contentPath' exists! (not copying the WordPress version)"
			sourceTarArgs+=( --exclude "./$contentPath" )
		fi
	done
	tar "${sourceTarArgs[@]}" . | tar "${targetTarArgs[@]}"
	echo >&2 "Complete! WordPress has been successfully copied to $PWD"
	if [ ! -e .htaccess ]; then
		# NOTE: The "Indexes" option is disabled in the php:apache base image
		cat > .htaccess <<-'EOF'
			# BEGIN WordPress
			<IfModule mod_rewrite.c>
			RewriteEngine On
			RewriteBase /
			RewriteRule ^index\.php$ - [L]
			RewriteCond %{REQUEST_FILENAME} !-f
			RewriteCond %{REQUEST_FILENAME} !-d
			RewriteRule . /index.php [L]
			</IfModule>
			# END WordPress
		EOF
		chown "www-data:www-data" .htaccess
	fi

	wp config create --allow-root --dbname=$WORDPRESS_DB_NAME --dbuser=$WORDPRESS_DB_USER --dbpass=$WORDPRESS_DB_PASSWORD --dbhost=$WORDPRESS_DB_HOST --dbcharset="utf8" --force --skip-check

	wp config set WP_DEBUG true --raw --allow-root
	wp config set WP_DEBUG_LOG true --raw --allow-root
	wp config set WP_DEBUG_DISPLAY true --raw --allow-root
	wp config set FS_METHOD "'direct'" --raw --allow-root

	echo "Installing WP Core"
	wp core install --url="http://wclocal:8080" --title="EBANX Pay WooCommerce Gateway" --admin_user=$EBANX_ADMIN_USERNAME --admin_password=$EBANX_ADMIN_PASSWORD --admin_email="plugins@ebanxpay.com" --allow-root

	echo "Install and activate storefrontheme"
	wp theme install storefront --version=$EBANX_STOREFRONT_THEME_VERSION --activate --allow-root

	echo "Install and activate plugins"
	wp plugin install woocommerce --version=$WOOCOMMERCE_PLUGIN_VERSION --activate --allow-root
	wp plugin activate woocommerce-gateway-ebanx-local --allow-root

	echo "Install Pages"
	wp post create --post_type=page --post_title='My Account' --post_status='publish' --post_content='[woocommerce_my_account]' --allow-root
	wp post create --post_type=page --post_title='Cart' --post_status='publish' --post_content='[woocommerce_cart]' --allow-root
	wp post create --post_type=page --post_title='Checkout' --post_status='publish' --post_content='[woocommerce_checkout]' --allow-root
	wp post create --post_type=page --post_title='Shop' --post_status='publish' --allow-root

	echo "Configure WooCommerce settings"
	wp db query 'UPDATE wp_options SET option_value="BR" WHERE option_name="woocommerce_default_country"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="BRL" WHERE option_name="woocommerce_currency"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="4" WHERE option_name="woocommerce_myaccount_page_id"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="5" WHERE option_name="woocommerce_cart_page_id"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="6" WHERE option_name="woocommerce_checkout_page_id"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="7" WHERE option_name="woocommerce_shop_page_id"' --allow-root
	wp db query 'UPDATE wp_options SET option_value="a:57:{s:19:\"sandbox_private_key\";s:30:\"test_ik_XRsjfeba9c8ibhVv10NUiw\";s:18:\"sandbox_public_key\";s:30:\"test_pk_ORbz0mIejNrVMrxvlUKhUQ\";s:20:\"sandbox_mode_enabled\";s:3:\"yes\";s:13:\"debug_enabled\";s:2:\"no\";s:22:\"brazil_payment_methods\";a:2:{i:0;s:20:\"ebanx-credit-card-br\";i:1;s:20:\"ebanx-banking-ticket\";}s:14:\"save_card_data\";s:3:\"yes\";s:9:\"one_click\";s:3:\"yes\";s:15:\"capture_enabled\";s:3:\"yes\";s:26:\"br_credit_card_instalments\";s:1:\"6\";s:13:\"due_date_days\";s:1:\"3\";s:20:\"brazil_taxes_options\";a:1:{i:0;s:3:\"cpf\";}s:25:\"br_interest_rates_enabled\";s:2:\"no\";s:17:\"show_local_amount\";s:3:\"yes\";s:27:\"br_min_instalment_value_brl\";s:1:\"5\";s:16:\"live_private_key\";s:0:\"\";s:15:\"live_public_key\";s:0:\"\";s:20:\"br_interest_rates_01\";s:0:\"\";s:20:\"br_interest_rates_02\";s:0:\"\";s:20:\"br_interest_rates_03\";s:0:\"\";s:20:\"br_interest_rates_04\";s:0:\"\";s:20:\"br_interest_rates_05\";s:0:\"\";s:20:\"br_interest_rates_06\";s:0:\"\";s:20:\"br_interest_rates_07\";s:0:\"\";s:20:\"br_interest_rates_08\";s:0:\"\";s:20:\"br_interest_rates_09\";s:0:\"\";s:20:\"br_interest_rates_10\";s:0:\"\";s:20:\"br_interest_rates_11\";s:0:\"\";s:20:\"br_interest_rates_12\";s:0:\"\";s:20:\"br_interest_rates_13\";s:0:\"\";s:20:\"br_interest_rates_14\";s:0:\"\";s:20:\"br_interest_rates_15\";s:0:\"\";s:20:\"br_interest_rates_16\";s:0:\"\";s:20:\"br_interest_rates_17\";s:0:\"\";s:20:\"br_interest_rates_18\";s:0:\"\";s:20:\"br_interest_rates_19\";s:0:\"\";s:20:\"br_interest_rates_20\";s:0:\"\";s:20:\"br_interest_rates_21\";s:0:\"\";s:20:\"br_interest_rates_22\";s:0:\"\";s:20:\"br_interest_rates_23\";s:0:\"\";s:20:\"br_interest_rates_24\";s:0:\"\";s:20:\"br_interest_rates_25\";s:0:\"\";s:20:\"br_interest_rates_26\";s:0:\"\";s:20:\"br_interest_rates_27\";s:0:\"\";s:20:\"br_interest_rates_28\";s:0:\"\";s:20:\"br_interest_rates_29\";s:0:\"\";s:20:\"br_interest_rates_30\";s:0:\"\";s:20:\"br_interest_rates_31\";s:0:\"\";s:20:\"br_interest_rates_32\";s:0:\"\";s:20:\"br_interest_rates_33\";s:0:\"\";s:20:\"br_interest_rates_34\";s:0:\"\";s:20:\"br_interest_rates_35\";s:0:\"\";s:20:\"br_interest_rates_36\";s:0:\"\";s:24:\"checkout_manager_enabled\";s:2:\"no\";s:35:\"checkout_manager_brazil_person_type\";s:0:\"\";s:27:\"checkout_manager_cpf_brazil\";s:0:\"\";s:28:\"checkout_manager_cnpj_brazil\";s:0:\"\";s:21:\"manual_review_enabled\";s:2:\"no\";}" WHERE option_name="woocommerce_ebanx-global_settings"' --allow-root

	echo "Create a product"
	wp wc product create --name='Jeans' --status='publish' --regular_price='250' --user=1 --allow-root

	echo "Configure Permalink"
	wp rewrite structure '/%postname%/' --hard --allow-root
fi

cd /var/www/html/wp-content/plugins/woocommerce-gateway-ebanx-local && composer install

echo "EBANX: Visit http://wclocal:8080/shop/ or http://wclocal:8080/wp-admin/"
echo "EBANX: Username - $EBANX_ADMIN_USERNAME"
echo "EBANX: Password - $EBANX_ADMIN_PASSWORD"

exec "$@"
